File: BE/src/main/java/udpm/hn/server/core/admin/role/service/RoleService.java
package udpm.hn.server.core.admin.role.service;

import udpm.hn.server.core.admin.role.model.request.RoleRequest;
import udpm.hn.server.core.admin.role.model.response.PermissionResponse;
import udpm.hn.server.core.admin.role.model.response.RoleResponse;

import java.util.List;
import java.util.Set;

public interface RoleService {

    /**
     * Get all roles
     */
    List<RoleResponse> getAllRoles();

    /**
     * Get role by ID with permissions
     */
    RoleResponse getRoleById(String id);

    /**
     * Create new role
     */
    RoleResponse createRole(RoleRequest request);

    /**
     * Update existing role
     */
    RoleResponse updateRole(String id, RoleRequest request);

    /**
     * Delete role
     */
    void deleteRole(String id);

    /**
     * Assign permissions to role
     */
    RoleResponse assignPermissionsToRole(String roleId, Set<String> permissionIds);

    /**
     * Get all permissions available in the system
     */
    List<PermissionResponse> getAllPermissions();

    /**
     * Get user's aggregated permissions from all their roles
     */
    Set<String> getUserPermissions(String userId);

    /**
     * Check if user has specific permission
     */
    boolean hasPermission(String userId, String permissionCode);
}

File: FE/src/services/admin/role.service.ts
import request from '../request'

export interface PermissionResponse {
  id: string
  name: string
  code: string
  description: string
  category: string
}

export interface RoleResponse {
  id: string
  name: string
  description: string
  permissions: PermissionResponse[]
  createdAt: number
  updatedAt: number
}

export interface RoleRequest {
  name: string
  description?: string
  permissionIds?: string[]
}

const roleService = {
  // Get all roles
  getAllRoles: async (): Promise<RoleResponse[]> => {
    const response = await request.get('/api/admin/roles')
    return response.data
  },

  // Get role by ID
  getRoleById: async (id: string): Promise<RoleResponse> => {
    const response = await request.get(`/api/admin/roles/${id}`)
    return response.data
  },

  // Create new role
  createRole: async (data: RoleRequest): Promise<RoleResponse> => {
    const response = await request.post('/api/admin/roles', data)
    return response.data
  },

  // Update role
  updateRole: async (id: string, data: RoleRequest): Promise<RoleResponse> => {
    const response = await request.put(`/api/admin/roles/${id}`, data)
    return response.data
  },

  // Delete role
  deleteRole: async (id: string): Promise<void> => {
    await request.delete(`/api/admin/roles/${id}`)
  },

  // Assign permissions to role
  assignPermissions: async (roleId: string, permissionIds: string[]): Promise<RoleResponse> => {
    const response = await request.put(`/api/admin/roles/${roleId}/permissions`, permissionIds)
    return response.data
  },

  // Get all permissions
  getAllPermissions: async (): Promise<PermissionResponse[]> => {
    const response = await request.get('/api/admin/roles/permissions')
    return response.data
  },
}

export default roleService
